### ISSUE TITLE
NameError vs. UnboundLocalError with walrus

### ISSUE THREAD

**Issue by JelleZijlstra** [carljm/compgenerator#4]
The fuzzer found this (slightly simplified):

```
def a(b):
    [(1) for b in b if (a := a)]
a("a")
```

On main:
```
>>> a("a")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in a
UnboundLocalError: cannot access local variable 'a' where it is not associated with a value
```

On 3.11:

```
>>> a("a")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in a
  File "<stdin>", line 2, in <listcomp>
NameError: cannot access free variable 'a' where it is not associated with a value in enclosing scope
```

I think that's expected: previously a walrus like this would create an implicit nonlocal, but now it's just a local. Perhaps this should be documented somewhere, though it's hard to imagine real code being affected.

For the fuzzer, we should somehow ignore this difference; I'll try to think of a way.

**Comment by JelleZijlstra** [comment]
Similarly, the qualname of lambdas defined inside a listcomp changes:

```
E         {'__a': {'__a': '<function __a.<locals>.<listcomp>.<lambda> at 0x...>', '__b': "'arg0'"}} != {'__a': {'__a': '<function __a.<locals>.<lambda> at 0x...>', '__b': "'arg0'"}}
```

**Comment by carljm** [comment]
It's neat to see the fuzzer actually working and finding things :)

Since `UnboundLocalError` is a subclass of `NameError`, I'm not sure this even rises to the level of needing mention in news? But happy to add an entry if you think that'd be best.

For ignoring this, we could use a regex to extract the variable name from `NameError` or `UnboundLocalError` and then simplify the representation of both errors to something simpler like "NameError: a".

Though that would also ignore a case where we changed an `UnboundLocalError` to a `NameError` -- I don't _think_ 709 should ever do that, but I would want the fuzzer to find it if we did.

I guess we could transform all `NameError` to the simpler error message form, but only transform `UnboundLocalError` to the simpler form of `NameError` on the remote evalserver. Then we'd catch it in one direction but ignore it in the other.

I think the qualname change should be ignored; we should be able to just strip `.<listcomp>` from qualnames to do that?
